openapi/logger.go

```diff
@@ -0,0 +1,52 @@
+package openapi
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/labstack/echo/v4"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// ZapLogger is a middleware and zap to provide an "access log" like logging for each request.
+func ZapLogger(log *zap.Logger) echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			start := time.Now()
+
+			err := next(c)
+			if err != nil {
+				c.Error(err)
+			}
+
+			req := c.Request()
+			res := c.Response()
+
+			fields := []zapcore.Field{
+				zap.Int64("size", res.Size),
+				zap.String("host", req.Host),
+				zap.Int("status", res.Status),
+				zap.String("method", req.Method),
+				zap.String("protocol", req.Proto),
+				zap.String("remote_ip", c.RealIP()),
+				zap.String("user_agent", req.UserAgent()),
+				zap.String("duration", time.Since(start).String()),
+				zap.String("request", fmt.Sprintf("%s %s", req.Method, req.RequestURI)),
+			}
+
+			n := res.Status
+			switch {
+			case n >= 500:
+				log.With(zap.Error(err)).Error("Server error", fields...)
```

Comment:
do we still need to print the error? line 20 has called it

Reply:
yes, we need, because L20 is actually call `terrorHTTPErrorHandler`, and we still need record ths http 5xx in dm-master's log